File: lld/code1.cpp
#include <iostream>
#include <string>

using namespace std;

class Car {
public:
    virtual void startEngine() = 0;
    virtual void shiftCar(int gear) = 0;
    virtual void brake() = 0;
    virtual void stopEngine() = 0;
    virtual ~Car() {}   // virtual destructor
};

class SportsCar : public Car {
public:
    string brand;
    string model;
    bool isEngine;
    int currentSpeed;
    int currentGear;

    SportsCar(string b, string m) {
        brand = b;
        model = m;
        isEngine = false;
        currentSpeed = 0;
        currentGear = 0;
    }

    void startEngine() override {
        cout << "Engine started\n";
        isEngine = true;
    }

    void shiftCar(int gear) override {
        currentGear = gear;
        cout << "Gear shifted to " << gear << endl;
    }

    void brake() override {
        cout << "Brake applied\n";
    }

    void stopEngine() override {
        cout << "Engine stopped\n";
        isEngine = false;
    }
};

int main() {
    Car* car = new SportsCar("Ford", "Mustang");
    car->startEngine();

    delete car; // VERY important
    return 0;
}

File: lld/code1.java
package lld;

abstract class Car {

    abstract void startEngine();
    abstract void shiftCar(int gear);
    abstract void brake();
    abstract void stopEngine();
}

class SportsCar extends Car {


   private String brand;
  private  String model;
   private boolean isEngine;
  private  int currentSpeed;
 private   int currentGear;

    SportsCar(String brand, String model) {
        this.brand = brand;
        this.model = model;
        this.isEngine = false;
        this.currentSpeed = 0;
        this.currentGear = 0;
    }

public String getBrand() {
        return brand;
    }

    public String getModel() {
        return model;
    }

    public boolean isEngineOn() {
        return isEngine;
    }

    public int getCurrentSpeed() {
        return currentSpeed;
    }

    public int getCurrentGear() {
        return currentGear;
    }

    // 🔹 Setters
    public void setCurrentGear(int gear) {
        if (gear >= 0 && gear <= 6) {
            this.currentGear = gear;
        }
    }

    public void setCurrentSpeed(int speed) {
        if (speed >= 0) {
            this.currentSpeed = speed;
        }
    }

    @Override
    void startEngine() {
        System.out.println("Engine started");
        isEngine = true;
    }

    @Override
    void shiftCar(int gear) {
        currentGear = gear;
        System.out.println("Gear shifted to " + gear);
    }

    @Override
    void brake() {
        System.out.println("Brake applied");
    }

    @Override
    void stopEngine() {
        System.out.println("Engine stopped");
        isEngine = false;
    }
}

public class code1 {
    public static void main(String[] args) {

        Car car = new SportsCar("Ford", "Mustang");
        car.startEngine();
        car.shiftCar(5);
     
    }
}
